### ISSUE TITLE
Use single planet services model for planetary UI actions

### ISSUE THREAD

**Issue by petervdmeer** [endless-sky/endless-sky#4530]
**Is your feature request related to a problem? Please describe.**
Plugins can use data-files to perform cosmetic changes to the Endless Sky user-interface. The feature from this change request should also allow multiple planetary screens to be completely redefined by just using data-files.

**Related Issue Links**
- Changes like #4516 should be implementable with some data-file changes and a lot less source-code changes (preferably no code changes and only datafile changes).
- This change brings us closer to "More generic UI code to make it easier to alter the user interface." from https://github.com/endless-sky/endless-sky/wiki/DevelopmentRoadmap
- The feature from this change would become very powerful if shop-panels can also be configured from datafiles (as described in #4529).

**Describe the solution you'd like**
- Using a single "planet services" model instead of separate models like shipyard-model, outfitter-model, trading-panel-model, hiring-model.
- Allow UI panels to start other UI panels.
  - Buttons should not just be used to communicate to the model directly, but should also be usable to open another view.

**Describe alternatives you've considered**
Also adding a universe model, but also adding that seems a bit to much for a single PR. Such universe model would work like:
- Planet model would support every action that can be done on the planet, like buying/selling ships and outfits and crew
- Universe model would support every action that can be done anywhere in a paused game, like setting a route or selecting ships

**Additional context**
Endless Sky uses a model-view based user interface, this change would allow multiple views on a single model.

**Comment by tehhowch** [comment]
Allowing an interface definition to specify the panel opened by a button (within reason) is something I've mulled over, which seems similar to the panel -> panel bit you mention.

The UI was one of the first bits of the game that came together, and it definitely shows.

**Comment by petervdmeer** [comment]
> Allowing an interface definition to specify the panel opened by a button (within reason) is something I've mulled over, which seems similar to the panel -> panel bit you mention.

Yes, that is exactly what I meant in the "Allow UI panels to start other UI panels" section.

Is it within reason if the shipyard-pannel and outfitter-pannel only exist in the datafiles, not in the source-code?

**Comment by tehhowch** [comment]
By "within reason" I mean a panel could only initialize a different panel that requires at most all the information that that panel has access to.

Defining a panel to be opened by a UI button that would require unavailable information would be an identifiable load-time user error.

**Comment by petervdmeer** [comment]
> By "within reason" I mean a panel could only initialize a different panel that requires at most all the information that that panel has access to.

Agreed, a planet panel can open a shipyard panel, but a planet panel should not open a boarding panel. That's what I meant with the 'single "planet services" model', panels that use planet services should only open other panels that use planet services (or universal services).

> Defining a panel to be opened by a UI button that would require unavailable information would be an identifiable load-time user error.

No, not an user-error, but a content bug. Panels are defined by content creators and plugin creators.

I was expecting to find such issues as run-time issue (panel trying to access unavailable information, resulting in an error-message or in a panel with empty fields, or both). Can you share a bit more on how to identify them during load-time?

**Comment by tehhowch** [comment]
In this context, the user in "user error" refers to whomever made the change to the given interface.

When parsing the interface, we know the name of the parsed node, and thus can determine which other interfaces are accessible from it, and can throw the appropriate runtime error.
